File: 2014-2015/Metody Programowania/Rozwiązania zadań/Lista 03/zad5.pl
insert([],H,[H]).
insert([H|L],Elem, Res):-
	Elem=<H, !,
	Res=[Elem,H|L].
insert([H|L],Elem, Res):-
	insert(L,Elem,Res1),
	Res=[H|Res1].

ins_sort([],[]).
ins_sort([H|L],S):-
	ins_sort(L,S1),
	insert(S1,H,S).

File: 2014-2015/Lato/Metody Programowania/Rozwiązania zadań/Lista 12/zad5.hs
data Cyclist a = Elem (Cyclist a) a (Cyclist a)

fromList :: [a] -> Cyclist a

fromList (x:xs) = res
	where
		res = Elem prev x next
		(next, prev) = aux res xs res

aux prev [] first = (first, prev)
aux prev (x:xs) first = (this, last)
	where
		this = Elem prev x tmp
		(tmp, last) = aux this xs first

-- Ponizej : Bardzo latwe !

forward :: Cyclist a -> Cyclist a
forward ( Elem _ _ a  ) = a

backward :: Cyclist a -> Cyclist a
backward ( Elem a _ _ ) = a

label :: Cyclist a -> a
label ( Elem _ a _ ) = a

--label.forward.forward.forward.backward.forward.forward $ fromList [1,2,3] = 2
